File: frontend/src/types/accessControl.ts
export type AccessDeviceStatus = "ONLINE" | "OFFLINE" | "MAINTENANCE";
export type FacialTemplateSyncStatus = "PENDING" | "SYNCED" | "FAILED";

export interface AccessDevice {
  id: string;
  name: string;
  location?: string | null;
  status: AccessDeviceStatus;
  last_seen_at?: string | null;
  created_by_id: string;
  created_at: string;
  updated_at: string;
}

export interface AccessDeviceWithKey extends AccessDevice {
  device_key: string;
}

export interface AccessDeviceCreate {
  name: string;
  location?: string | null;
}

export interface AccessDeviceStatusUpdate {
  status: AccessDeviceStatus;
}

export interface PaginatedAccessDeviceRead {
  items: AccessDevice[];
  total: number;
}

export interface FacialTemplate {
  id: string;
  resident_id: string;
  media_asset_id: string;
  sync_status: FacialTemplateSyncStatus;
  synced_at?: string | null;
  failure_reason?: string | null;
  created_at: string;
  updated_at: string;
}

export interface FacialAccessEvent {
  id: string;
  device_id: string;
  resident_id?: string | null;
  matched: boolean;
  confidence_score?: number | null;
  access_granted: boolean;
  event_time: string;
  raw_payload?: string | null;
}

export interface PaginatedFacialAccessEventRead {
  items: FacialAccessEvent[];
  total: number;
  skip: number;
  limit: number;
}

File: backend/app/schemas/access_control.py
"""Pydantic schemas for AccessDevice, FacialTemplate, and FacialAccessEvent."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.models.enums import AccessDeviceStatus, FacialTemplateSyncStatus


class AccessDeviceCreate(BaseModel):
    name: str
    location: str | None = None


class AccessDeviceStatusUpdate(BaseModel):
    status: AccessDeviceStatus


class AccessDeviceRead(BaseModel):
    id: UUID
    name: str
    location: str | None = None
    status: AccessDeviceStatus
    last_seen_at: datetime | None = None
    created_by_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AccessDeviceKeyRead(AccessDeviceRead):
    """Extends AccessDeviceRead to expose the secret key exactly once (create/regenerate)."""

    device_key: str


class PaginatedAccessDeviceRead(BaseModel):
    items: list[AccessDeviceRead]
    total: int


class FacialTemplateRead(BaseModel):
    id: UUID
    resident_id: UUID
    media_asset_id: UUID
    sync_status: FacialTemplateSyncStatus
    synced_at: datetime | None = None
    failure_reason: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class FacialVerificationWebhookIn(BaseModel):
    resident_id: UUID | None = None
    confidence_score: float | None = None


class FacialAccessEventRead(BaseModel):
    id: UUID
    device_id: UUID
    resident_id: UUID | None = None
    matched: bool
    confidence_score: float | None = None
    access_granted: bool
    event_time: datetime
    raw_payload: str | None = None

    model_config = ConfigDict(from_attributes=True)


class PaginatedFacialAccessEventRead(BaseModel):
    items: list[FacialAccessEventRead]
    total: int
    skip: int
    limit: int
